### ISSUE TITLE
Expose design-time type information in TC39 decorator metadata when `emitDecoratorMetadata: true`

### ISSUE THREAD

**Issue by artberri** [microsoft/TypeScript#57533]
### 🔍 Search Terms

decorators, metadata, emitDecoratorMetadata, "TC39 decorators", "type information", reflect-metadata

### ✅ Viability Checklist

- [X] This wouldn't be a breaking change in existing TypeScript/JavaScript code
- [X] This wouldn't change the runtime behavior of existing JavaScript code
- [X] This could be implemented without emitting different JS based on the types of the expressions
- [X] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
- [X] This isn't a request to add a new utility type: https://github.com/microsoft/TypeScript/wiki/No-New-Utility-Types
- [X] This feature would agree with the rest of our Design Goals: https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals

### ⭐ Suggestion

I want to be able to have design-time type information at runtime after the implementation of the new TC39 decorators system.

After the implementation of the [TC39 decorators](https://devblogs.microsoft.com/typescript/announcing-typescript-5-0/#decorators) in Typescript 5.0 the ability to emit design-time type information has been lost unless we keep using the experimental (legacy) decorators system.

The suggestion is to allow to use `emitDecoratorMetadata: true` with `experimentalDecorators: false`. The idea is to expose the same metadata that it was being exposed to `Reflect.metadata("design:type", type)` with the legacy decorators, to the [new metadata system proposed by TC39](https://github.com/tc39/proposal-decorator-metadata) that has been implemented in [Typescript 5.2](https://devblogs.microsoft.com/typescript/announcing-typescript-5-2/#decorator-metadata).

So, instead of using the obsolete `reflect-metadata` like `Reflect.getMetadata('design:type', target, property)`, we can get the same info by doing something like `SomeClass[Symbol.metadata]["design:type"]`.

### 📃 Motivating Example

After the decorators change, many existing and widely used libraries will have problems implementing their behaviors without type information at runtime: [type-graphql](https://github.com/MichalLytek/type-graphql), [inversify](https://inversify.io/), Angular 2,... or any other library which depends on `reflect-metadata`.

### 💻 Use Cases

As explained in the "Motivating Example", there are several existing use cases that require some kind of design-time type information at runtime. My suggestion is to use more or less the same approach you used with the experimental decorators, but with the standard ones. But I will mostly agree to any other approach that makes this information available at runtime.

For now, the workaround is to keep using the experimental decorators and the `reflect-metadata` library (or compatible). But I think it is worth giving the standard ones the same ability to expose this info.

**Comment by RyanCavanaugh** [comment]
>  This could be implemented without emitting different JS based on the types of the expressions

Can you give some examples of how this would work while maintaining this invariant?

**Comment by bcheidemann** [comment]
I've done some investigation into how this could be implemented. As far as I can tell, this would be a fairly trivial change to implement. I was able to get it working for a few basic examples with some pretty minimal changes:

```patch
diff --git a/src/compiler/factory/emitHelpers.ts b/src/compiler/factory/emitHelpers.ts
index 0937915e75..9e6e06a1d5 100644
--- a/src/compiler/factory/emitHelpers.ts
+++ b/src/compiler/factory/emitHelpers.ts
@@ -736,7 +736,9 @@ export const metadataHelper: UnscopedEmitHelper = {
     priority: 3,
     text: `
             var __metadata = (this && this.__metadata) || function (k, v) {
-                if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
+                return function (_, c) {
+                    c.metadata[k] = v;
+                }
             };`,
 };
 
diff --git a/src/compiler/program.ts b/src/compiler/program.ts
index c46e910a53..dd9f29cb1c 100644
--- a/src/compiler/program.ts
+++ b/src/compiler/program.ts
@@ -4357,13 +4357,6 @@ export function createProgram(rootNamesOrOptions: readonly string[] | CreateProg
             }
         }
 
-        if (
-            options.emitDecoratorMetadata &&
-            !options.experimentalDecorators
-        ) {
-            createDiagnosticForOptionName(Diagnostics.Option_0_cannot_be_specified_without_specifying_option_1, "emitDecoratorMetadata", "experimentalDecorators");
-        }
-
         if (options.jsxFactory) {
             if (options.reactNamespace) {
                 createDiagnosticForOptionName(Diagnostics.Option_0_cannot_be_specified_with_option_1, "reactNamespace", "jsxFactory");
diff --git a/src/compiler/transformers/ts.ts b/src/compiler/transformers/ts.ts
index f7af3ceb8e..9f51a89ca7 100644
--- a/src/compiler/transformers/ts.ts
+++ b/src/compiler/transformers/ts.ts
@@ -1112,7 +1112,6 @@ export function transformTypeScript(context: TransformationContext) {
      */
     function getTypeMetadata(node: Declaration, container: ClassLikeDeclaration) {
         // Decorator metadata is not yet supported for ES decorators.
-        if (!legacyDecorators) return undefined;
         return USE_NEW_TYPE_METADATA_FORMAT ?
             getNewTypeMetadata(node, container) :
             getOldTypeMetadata(node, container);

```

Of course, in order to still support legacy decorators, some additional changes would be required. But these would likely also be straightforward.

**Comment by bcheidemann** [comment]
@RyanCavanaugh I've raised a draft PR to serve as a proposal for how this could be implemented, and to serve as a PoC. Please let me know if there's anything else I can do to help progress this issue :)

**Comment by fatcerberus** [comment]
> So, if this invariant [no different JS based on types] was met by the previous implementation, it would also be met by the new one.

fwiw the invariant *isn't* met by the legacy implementation - it falls under the grandfather clause because it was done before the "no type-directed emit" rule was instituted.  This is similar to how

> This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)

is violated by `enum` and `namespace`, for the same reasons.

**Comment by bcheidemann** [comment]
Thanks for clarifying @fatcerberus. Do you know what the thinking is on this... since it was supported for legacy decorators, is it likely to continue being supported with the new ES decorators in some capacity despite violating this invariant?

Also could you confirm how exactly the invariant is violated by this feature? It might be helpful to have some explanation of this if the feature ends up being rejected on these grounds.

**Comment by bruceharrison1984** [comment]
If this is rejected, is there some alternative proposal that would emerge? 

With Stage 3 metadata, the current lack of something such as `design:paramtypes` greatly reduces the utility of decorators for something like dependency injection as an obvious example. To further that point, most existing DI libraries could never transition to Stage 3 Decorators due to the lack of emitted metadata and would be stuck with the `experimentalDecorator` option.

**Comment by samueldcorbin** [comment]
Moreover, that split probably harms the project goals even more than just adding this feature.

It feels like this is being treated as a choice between violating that goal and not violating it.

But since this feature is in widespread use and the behavior can't be replicated without it, the actual choice seems to be between having a violation of that goal and having a violation of that goal _plus_ an awkward, non-conforming dual-path decorator spec.

**Comment by RomainLanz** [comment]
Hi everyone! 👋🏻

Are there any updates on a potential migration path for dependency injection?

When using the official decorator proposal, the TypeScript compiler doesn't emit type metadata as it did with the legacy version.

```ts
function inject() {
  return function(target: any, context: any): void {
    console.log(context) // Metadata is empty
  }
}

class Service {}

@inject()
class Controller {
  constructor(private service: Service) {}
}
```

```json
{
  "kind": "class",
  "name": "Controller",
  "metadata": {}
} 
```

In this example, I would expect the metadata field to include `Service` when `emitDecoratorMetadata` is set to `true`. Maybe we could introduce a new flag specifically for this purpose.

This is the only missing piece needed to migrate away from `reflect-metadata` for many frameworks and libraries.

**Comment by RomainLanz** [comment]
If official ESM decorators aren't going to emit type metadata, now or in the future, knowing that would help us make more informed decisions about whether to migrate away from legacy decorators and `reflect-metadata`.

Right now, we risk wasting time implementing workarounds or alternative syntaxes, only for TypeScript to support this feature later. Conversely, if type metadata isn't on the roadmap at all, **that’s something the community needs to know as soon as possible**.

**Could the TypeScript team please share any clarity about the likelihood and timeline (or lack thereof) for supporting design-time type metadata emission under ESM decorators?**

That information alone would go a long way toward helping frameworks and libraries chart their path forward. Thank you!

**Comment by zangopushpak** [comment]
Are there any updates on a potential migration path for dependency injection?

**Comment by bcheidemann** [comment]
> Are there any updates on a potential migration path for dependency injection?

@zangopushpak there are many DI solutions which do not depend on design-time type information. Depending on your project, you may or may not be able to migrate to one of these. In my opinion, [typed-inject](https://www.npmjs.com/package/typed-inject) is an excellent option, as it provides type-safety with a reasonably ergonomic API.

**Comment by RomainLanz** [comment]
> > Are there any updates on a potential migration path for dependency injection?
> 
> [@zangopushpak](https://github.com/zangopushpak) there are many DI solutions which do not depend on design-time type information being available at runtime. Depending on your project, you may or may not be able to migrate to one of these. In my opinion, [typed-inject](https://www.npmjs.com/package/typed-inject) is an excellent option, as it provides type-safety with a reasonably ergonomic API.

That’s not really the point. We already know how to implement DI without runtime type information. The issue is whether we need to migrate all our codebases to an alternative approach, or if TypeScript plans on providing a path for ESM decorators to emit the necessary metadata.

**We just need a straight answer so we can decide whether to wait or start rewriting everything.**

**Comment by thomas-darling** [comment]
I also want to point out that we are at the mercy of the frameworks we use. Just because something could _theoretically_ be done in a different way, doesn't mean we can just magically migrate to doing it that way. This is not a decision we can make, but a decision to be made by the maintainers of the many frameworks and dependencies we've been relying on for a decade.

Also, things like `typed-inject` is not, in fact, a viable replacement for this, as all these other solutions require lots of redundant boilerplate code to register things and declare dependencies in specific ways. The entire point of using the metadata is to avoid that mess, which is why so much code and so many frameworks have come to rely so heavily on it.

Personally, I'm going to be **majorly** pissed off if this ends up not being supported, as that means we will have several very large and business critical codebases that we will simply never be able to update to a new TypeScript version - just because of this. Breaking fundamental things that have been supported for this long, with so much code and business depending on it, is simply not acceptable, regardless of whatever project goals TypeScript may have.

Like it or not, you implemented an extremely useful feature, and now your customers business depend on it.

**Therefore, I don't just need a straight answer, but a firm commitment to supporting this.**

---

As for maintaining the invariant mentioned by @RyanCavanaugh at the beginning of this discussion, my answer is this:

_This feature does not cause **different** JS to be emitted - it is an **opt-in** that causes **additional** JS to be emitted._

**Comment by RyanCavanaugh** [comment]
I don't think we're going to have decorator metadata in ES decorators. It's unlikely. Not impossible, but very unlikely. We are increasingly moving away from any type-based emit in all forms; it's slow, error-prone, impossible for third-party tools to replicate, and not standardized.
 
`emitDecoratorMetadata` is documented as experimental, and also requires you to set another flag with `experimental` in the name. Features don't stop becoming experimental due to the passage of time -- if we thought they were not experimental, we would have renamed them. They're still experimental, and it's still up to each developer to understand the consequences of taking dependencies on features that are not considered stable.
 
We tried for a very long time to align the TC39 decorator proposal with the experimental decorators that originated in TypeScript. The committee didn't land on something that exactly aligns with what we have, with one delta being the lack of metadata. It's possible the committee will eventually update decorators to include that, but that's a decision for the committee to make, not ours. The reason you saw an experimental feature in TS for so long is because we were trying so hard to make it *not* experimental, by driving our proposal through the committee, but a) committees take a long time and b) where we landed in the committee doesn't support this feature, for better or worse.

Our hope is that most experimental decorators users will be able to transition to ES decorators, but they are definitely not at 100% parity and reaching forcing that parity by implementing another nonstandard feature just digs the TS/JS divergence hole deeper, and that runs contrary to our goal of getting the ecosystem standardized on one set of runtime semantics.

**Comment by bcheidemann** [comment]
@RyanCavanaugh thanks for clarifying the teams stance on this feature. Myself and others have made our thoughts clear in the above discussion, but I recognise and respect the TS teams right to maintain this project however they see fit. I'm very grateful for all the amazing work you and others do to benefit all of us 🙂

**Comment by thetutlage** [comment]
The similar takes were taken on using `.js` file extension for importing TypeScript files. I debated about it in many issues and they all were shutdown with the same argument of TypeScript philosophy.

I hope some influential projects feeling the heat of this decision will let you guys re-think about it in the future.

**Comment by RyanCavanaugh** [comment]
I don't know how to convince you that a world where we abandon those principles is definitely worse, but it is. TypeScript got to where it is because of those good principles, not despite them.

**Comment by bcheidemann** [comment]
> ... we will have several very large and business critical codebases that we will simply never be able to update to a new TypeScript version

@RyanCavanaugh To assuage the concerns of @thomas-darling (and presumably others), will the `experimentalDecorators` and `emitDecoratorMetadata` flags continue to be supported in future versions of TypeScript?

**Comment by nomagick** [comment]
I would like to remind you that TypeScript was only widely adopted after the merge with AtScript.

AtScript has the idea in its name. Decorator and run-time access to design-time information would bring a programming language forward and to the next level. Look at the adoption of type anonations in Python.

With the Microsoft team's current attitude towards the idea of "run-time type metadata" and Decorators, a fork of a new AtScript is very likely. Maybe I'll personally do it at some point.

No offence to anyone, but this long refusal to cooperate attitude appears to me that the Decorators stuff was like a forced deal with Google that you now regret.

Maybe TypeScript has done its part and is ready to retire, like CoffeeScript.
And the journey beyond compile-time type checking must be led by someone else. 
Perhaps this `Annotation + Introspection` perspective does need a separate name.

![Image](https://github.com/user-attachments/assets/08add545-0fad-4b0b-a710-b86174e3139c)

**Comment by nomagick** [comment]
TL;DR: If you are seeking to migrate existing framework that fully embraces Decorators, TC39 Decorators is not what you want. It's totally different and not really helpful.
No point migrating to it. What you need is a fork of TypeScript with a new name.

----

After checking the latest status of several TC39 proposals, I found that we are still at the very early stages of standardizing the Decorators as we know it in TypeScript.

https://github.com/tc39/proposal-decorators
https://github.com/tc39/proposal-decorator-metadata

We have Decorators and Decorator metadata proposal in stage 3, which initially seems promising.
However, the decorator proposal did not include "parameter decorators", which is important for the current Decorator based ecosystem to fully work.

https://github.com/tc39/proposal-class-method-parameter-decorators

Note this proposal is in stage 1 proposed in March 2023.

If you dive into the technical details of the proposals, you'll find not only that the proposal is at early stages, but also that the whole concept of Decorators is fundamentally different from the one in TypeScript.

It appears to me that it's impossible for TC39 method/property decorators to get a reference to the Class without instantiating the class.
It seems impossible to me to properly register every piece of metadata like the way it is in current TypeScript Decorators.

So we are effectively at stage 0 now, one decade after the set of features being widely adopted.

There's clearly some dysfunction happening in TC39.

So indeed it's not really helpful to merge any of the PRs trying to adopt TC39 decorators with the intention to "migrate" existing code.

I take back my previous statement that TypeScript seems to be retiring. 
But it's still true that TypeScript is not pushing in this direction.
It's like Decorators is some kind of bastard, and TypeScript is unwilling to embrace and adopt Decorators to be its own feature, marking it "experimental" indefinitely.

I guess indeed we need some kind of fork to make clear that the feature set of the previous [AtScript](https://en.wikipedia.org/wiki/AtScript) is still relevant and useful. Making it not "experimental" first as a start.

**Comment by mlhaufe** [comment]
@nomagick :

> It appears to me that it's impossible for TC39 method/property decorators to get a reference to the Class without instantiating the class.

```ts
const myDecorator= (_feature, ctx) => {
    // ...
    ctx.addInitializer(function () {
        const myClass = this

        // ...
    })
}
```

**Comment by nomagick** [comment]
@mlhaufe If I understand correctly, `this` in method/prop initializer is the instance of the class, and it runs when the class is being instantized.
https://github.com/tc39/proposal-decorators/tree/master?tab=readme-ov-file#example-bound

Bizarre enough, but I do understand the trick now https://github.com/tc39/proposal-decorators/tree/master?tab=readme-ov-file#example-bound. 
Despite you can't get the Class, it's still possible for one-way communication.
It's closure. Prop/method decorators can put data into a shared object that class decorators later access.
You need to first create something like a "decorators kit" every time, you destruct the kit and take the brand new dedorators it contains. Data is shared using this one-time kit.

**Comment by mlhaufe** [comment]
@nomagick 

> Despite you can't get the Class, it's still possible for one-way communication.

Not sure what you mean here. `this.constructor` in the initializer would be the class. This initializer is run after `super` and before field initialization. 

> Prop/method decorators can put data into a shared object that class decorators later access.

In other words the existing`ctx.metadata`

> You need to first create something like a "decorators kit" every time, you destruct the kit and take the brand new dedorators it contains. Data is shared using this one-time kit.

Can you expand on what this "decorators kit" expresses?

**Comment by nomagick** [comment]
@mlhaufe 

>Not sure what you mean here. this.constructor in the initializer would be the class. This initializer is run after super and before field initialization.

I mean method/prop decorators cannot see the class without creating an instance of the class.


>In other words the existing ctx.metadata

Thanks for mentioning this. ctx.metadata of each class is the same object across class/method/prop decorators.
This is indeed a better way to share data between decorators.

>Can you expand on what this "decorators kit" expresses?

I mean [this](https://github.com/tc39/proposal-decorators/blob/8936bfe642c63db0c6565572c4847c120400dac3/README.md?plain=1#L815) and [this](https://github.com/tc39/proposal-decorators/blob/8936bfe642c63db0c6565572c4847c120400dac3/README.md?plain=1#L775-L789). In the original Decorators proposal. It's already possible to do one-way data sharing by dynamically creating multiple Decorators in groups and sharing an object within the group. 

TC39 Decorators + metadata turned out to be legit.

I have learned a lot today from this issue. 
I am generally sorry for the disturbance. Forgive my ignorance

**Comment by mlhaufe** [comment]
@nomagick 

An excellent article on this has been writting by @rauschma you might enjoy:

https://2ality.com/2022/10/javascript-decorators.html

**Comment by zookatron** [comment]
> Our hope is that most experimental decorators users will be able to transition to ES decorators, but they are definitely not at 100% parity and reaching that parity by implementing another nonstandard feature just digs the TS/JS divergence hole deeper, and that runs contrary to our goal of getting the ecosystem standardized on one set of runtime semantics.

@RyanCavanaugh It seems to me like you're approaching this issue purely from the perspective of ECMAScript compatibility, and while that is a laudable goal, it seems to be missing the point of this discussion. This proposal is specifically about having the official TypeScript compiler emit static type metadata, which is not something that can ever possibly be directly compatible with ECMAScript itself because ECMAScript _doesn't have a notion of static types_. If static type annotations _were_ added to ECMAScript, it would essentially make TypeScript itself unnecessary, making not only this proposal but the entire static type ecosystem that Microsoft has cultivated largely irrelevant.

If this policy were sustained, it would make TypeScript a huge outlier among modern statically typed programming languages. Virtually every single language of the [top 50 used on GitHub](https://innovationgraph.github.com/global-metrics/programming-languages) that supports any type of static type information also supports runtime reflection of that data (Python, C++, Java, PHP, C#, Go, Swift, etc). There's a reason why so many major TypeScript projects rely on the emitDecoratorMetadata experimental feature, it's because the patterns that this type of reflection enables are incredibly common and useful across a wide variety of programming ecosystems.

As @nomagick mentions, if the TypeScript compiler itself did not support this functionality, it would essentially force developers to use an AtScript-like preprocessor that generates TypeScript that emits metadata to enable runtime reflection. This would complicate deployment and code reuse processes for untold thousands of developers across the world who need these features, and require the introduction of a new extension (maybe "*.tspp"? 😉) that code bundlers would have to support in order to enable the full static typing functionality that developers need.

I definitely understand the value of preserving the ease of adoption of the TypeScript language by trying as hard as possible to make it possible to parse TypeScript simply by removing type information, rather than requiring emitting additional JavaScript code during the parsing process. This has allowed tools like `esbuild` to support the language relatively easily, since they skip supporting the emitDecoratorMetadata feature. However, there's simply no other logical place for this particular functionality to be supported other than in the TypeScript compiler itself because no other layer actually has access to this static type information which needs to be emitted. Therefore, this desire for simplicity should not preclude the inclusion of static type metadata emission in the official TypeScript specification.

As someone who would love to see TypeScript become the lingua franca of the entire web development ecosystem, it would break my heart to see the ecosystem split into a legacy TypeScript ecosystem and a new "TypeScript++" language that actually supports runtime reflection of static types, especially when this functionality is already mostly implemented and just needs to be tweaked to support the new TC39 proposal, which many people here would be happy to assist with. Please help preserve the unity of the statically typed web development world by prioritizing this important functionality!
